apiserver/facades/client/charms/client.go

```diff
@@ -475,25 +475,27 @@ func (a *API) checkCharmPlacement(ctx context.Context, arg params.ApplicationCha
 
 	// Unfortunately we now have to check instance data for all units to
 	// validate that we have a homogeneous setup.
-	units, err := app.AllUnits()
-	if err != nil {
+	unitNames, err := a.applicationService.GetUnitNamesForApplication(ctx, arg.Application)
+	if errors.Is(err, applicationerrors.ApplicationNotFound) {
+		return params.ErrorResult{}, nil
+	} else if err != nil {
 		return params.ErrorResult{
 			Error: apiservererrors.ServerError(err),
 		}, nil
 	}
 
 	arches := set.NewStrings()
-	for _, unit := range units {
-		machineID, err := unit.AssignedMachineId()
-		if errors.Is(err, errors.NotAssigned) {
+	for _, unitName := range unitNames {
+		machineName, err := a.applicationService.GetUnitMachineName(ctx, unitName)
+		if errors.Is(err, applicationerrors.UnitMachineNotAssigned) {
```

Comment:
todo: combine into one call. The only reason to get the unit names to find the machines the units are on.

Ideally this entire block of code from 476 to 522 could be pushed into a domain.

Reply:
As discussed, I have extracted the bulk of of this logic into a stub service method on the machine domain. This should be implemented when we get back to the machine domain